### PR TITLE
Add error message for using >= 65535 hashes for raw string literal escapes

### LINKED ISSUE #50111: Raw string literal escape depth is technically limited
This is not entirely serious, because I can't imagine this plausibly happening in real-world code (generated or otherwise), but in Nightly (Stable has a higher threshold):
```rust
fn main() {
    r#[... repeat 2^16 times ...]#""#[... repeat 2^16 times ...]#;
}
```
fails to parse with:
```
error: expected one of `.`, `;`, `?`, `}`, or an operator, found `#`
 --> src/main.rs:3:65544
  |
3 |     r######...
  | 
```

This limit is not mentioned [in the reference](https://doc.rust-lang.org/reference/tokens.html#raw-string-literals).

I think we should document the maximum number of `#`s. We could also reduce the number permitted: 255 should be enough for anyone (https://github.com/rust-lang/rust/pull/49993#issuecomment-381630152).

Alternatively, we could ignore it as being a non-problem. Feel free just to close the issue 😄

### PR DESCRIPTION
**Pull request by dcao**
Fixes #50111.

### CODE REVIEW

File: src/test/ui/raw_string_hash_count.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// ignore-tidy-linelength
+
+fn main() {
+    let x = r################################################################################################################################################################################################################################################################"test"################################################################################################################################################################################################################################################################
```
Comment (varkor) [inline src/test/ui/raw_string_hash_count.rs]: For simplicity, you can eliminate the variable as well, and just have the raw string as a standalone statement.

File: src/libsyntax/parse/lexer/mod.rs [hunk 2]
```diff
@@ -1669,8 +1676,15 @@ impl<'a> StringReader<'a> {
...
         let start_bpos = self.pos;
         self.bump();
-        let mut hash_count = 0;
+        let mut hash_count: u8 = 0;
         while self.ch_is('#') {
+            if hash_count == 255 {
+                let last_bpos = self.pos;
+                self.fatal_span_(start_bpos,
+                                    last_bpos,
+                                    "too many `#` characters surrounding a raw string; \
```
Comment (varkor) [inline src/libsyntax/parse/lexer/mod.rs]: Maybe clarify to "raw byte string". It's probably fine to only test one of them, though, as it's a minor error. Also, the indentation here is off.

File: src/libsyntax/parse/lexer/mod.rs [hunk 3]
```diff
@@ -1439,8 +1439,15 @@ impl<'a> StringReader<'a> {
...
                 let start_bpos = self.pos;
                 self.bump();
-                let mut hash_count: u16 = 0;
+                let mut hash_count: u8 = 0;
                 while self.ch_is('#') {
+                    if hash_count == 255 {
+                        let last_bpos = self.pos;
+                        self.fatal_span_(start_bpos,
+                                         last_bpos,
+                                         "too many `#` characters surrounding a raw string; \
```
Comment (varkor) [inline src/libsyntax/parse/lexer/mod.rs]: The wording could be a little more concise here. Maybe:
```
too many `#` symbols: raw strings may be delimited by up to 255 `#` symbols
```
(and same for byte strings).

**Review by varkor — CHANGES_REQUESTED** [review]
It's looking good to me! I just left a few very minor comments.

The test is currently failing because the offset in the line/column number is wrong (after you added the line to ignore tidy), but once you tweak the range/text you can just regenerate it.

File: src/test/ui/raw_string_hash_count.stderr [hunk 4]
```diff
@@ -0,0 +1,8 @@
+error: too many `#` symbols: raw strings may be delimited by up to 255 `#` symbols
+  --> src/test/ui/raw_string_hash_count.rs:14:5
```
Comment (varkor) [inline src/test/ui/raw_string_hash_count.stderr]: You need to replace `src/test/ui` with `$DIR`.

File: src/test/ui/raw_string_hash_count.rs [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// ignore-tidy-linelength
+
+fn main() {
+    r################################################################################################################################################################################################################################################################"test"################################################################################################################################################################################################################################################################;
+    //~^ ERROR too many `#` characters
```
Comment (varkor) [inline src/test/ui/raw_string_hash_count.rs]: This needs to be updated to use `symbols` instead of `characters`.

**Comment by nagisa** [comment]
Probably for better diagnostics? We either introduce some limit or make the number of hashes technically unlimited by counting them into u64 or larger.

**Comment by nikomatsakis** [comment]
I feel ok about going to `u8` for now. [Searching through crates.io with a simple regex, it seems pretty clear that this will not break anything.][s]

[s]: https://sourcegraph.com/search?q=repogroup:crates+max:200+%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23%5C%23

**Comment by eddyb** [comment]
@nikomatsakis I don't thinks this helps anything. @nnethercote changed it to `u16` for a size win, but going to `u8` doesn't reduce that size any further.

**Comment by estebank** [comment]
@cmdd could you squash your commits?

**Comment by eddyb** [comment]
I'm nominating this PR for discussion, on the topic of what limits can an implementation reasonably have (rustc currently uses `u32` in a bunch of places where `usize` would be impossible to overflow).

**Comment by joshtriplett** [comment]
We discussed this in the lang team meeting, and we all agreed that we can safely limit this to 255 without concern.

**Comment by nikomatsakis** [comment]
I wasn't at the meeting. But I think @eddyb feels like in general `u8` is a rather small limit just *on principle*. I agree that it sounds like something you would expect in the 1970s, though I also agree that it ought be enough. =) If it brings no impl benefit over `u16`, let's just do that and call it a day.

**Comment by kennytm** [comment]
I do not want a 128 KB test case if `u16` is chosen  😄

**Comment by nnethercote** [comment]
I suggest 255. It's more than anyone will need in practice, but it's also small enough to test reasonably.

**Comment by pietroalbini** [comment]
Ping from triage @eddyb! Can we reach a consensus here?

**Comment by eddyb** [comment]
We don't test other limits (other than the ones imposed by LLVM), why should this be an exception?

**Comment by kennytm** [comment]
Ping again from not-a-triage @eddyb (reviewer) @cmdd (author) @rust-lang/lang (team), this PR is stuck. Does everybody agree with eddyb's [comment](https://github.com/rust-lang/rust/pull/50296#issuecomment-388528799) above? That is,

* We keep the type to be `u16` (almost reverts everything in this PR)
* Still add the check when the number of `#` reaches 65535 (similar to the implementation in this PR: https://github.com/rust-lang/rust/pull/50296/files#diff-d06ad31c547d2d94c8b8770006977767)
* *Do not* add any test cases

(Nominating again, feel free to untag and change to `S-waiting-on-author` when a decision is made.)

**Comment by varkor** [comment]
@kennytm: I think @eddyb was pushing not to handle this at all, in which case the PR should just be closed.

@cmdd: I'd be happy to mentor a less contentious issue if that's the case. Most of the [E-needstest](https://github.com/rust-lang/rust/labels/E-needstest) issues should be straightforward, as you just need to add a new test confirming that the original issue no longer occurs. Feel free to mention me on any if you'd like guidance.

**Comment by cmdd** [comment]
I think it makes sense to have an error message which covers the "too many hashes" case even if the number of hashes is defined as a u16, since it doesn't impede any behavior but provides a useful diagnostic for an (admittedly extreme) edge-case. If the final decision is to close the PR & the issue as "won't fix," please let me know; otherwise, I'll hopefully get this implemented soon and discussion can continue from there.

@varkor: Thanks for the mentoring help! Hopefully adding a test case won't result in such controversy :)

**Comment by cmdd** [comment]
I've kerjiggled with my branch so that the only change this PR makes is adding an error message when using 2^16 or more hashes to delimit a raw (byte) string. Hopefully this change is agreeable for everyone!

**Comment by pietroalbini** [comment]
Ping from triage @nikomatsakis! This PR needs your review.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Can we get a test ?

**Comment by nikomatsakis** [comment]
@cmdd I imagine adding a test into the `src/test/parse-fail` directory would make sense. You can read more about tests [in the rustc-guide](https://rust-lang-nursery.github.io/rustc-guide/tests/intro.html)

**Comment by kennytm** [comment]
@nikomatsakis But https://github.com/rust-lang/rust/pull/50296#issuecomment-388528799 doesn't want any tests 😕

**Comment by nikomatsakis** [comment]
I realize that -- I guess the thought is that 128k is a lot? didn't strike me as a lot. but fine, if @eddyb doesn't want tests, it's on their head. 😛 

@bors r+

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@cmdd merge conflict :(

**Comment by pietroalbini** [comment]
@cmdd you should rebase on master, even if you don't see any conflict at first glance.

**Comment by cmdd** [comment]
Sorry for the delay! I've rebased onto master.

**Pull request merged**